src/utils/directoryListing.ts

```diff
@@ -0,0 +1,119 @@
+import Handlebars from 'handlebars';
+import { File, ReadDirectoryResult } from '../providers/provider';
+import htmlTemplate from '../templates/directoryListing.out';
+import { toReadableBytes } from '../utils/object';
+
+const handlebarsTemplate = Handlebars.template(htmlTemplate);
+
+const months = [
+  'Jan',
+  'Feb',
+  'Mar',
+  'Apr',
+  'May',
+  'Jun',
+  'Jul',
+  'Aug',
+  'Sep',
+  'Oct',
+  'Nov',
+  'Dec',
+];
+
+/**
+ * Render a directory listing
+ *
+ * TODO: Work with other teams on removing their dependency on nginx's listing
+ *  result so we don't need to emulate it
+ */
+export function renderDirectoryListing(
+  url: URL,
+  result: ReadDirectoryResult
+): string {
+  const tableElements: TableElement[] = [];
+
+  for (const name of result.subdirectories) {
+    tableElements.push(renderSubdirectory(name));
+  }
+
+  let directoryLastModified: Date | undefined = undefined;
+  const urlPathname = `${url.pathname}${url.pathname.endsWith('/') ? '' : '/'}`;
+  for (const file of result.files) {
+    if (
+      directoryLastModified === undefined ||
+      file.lastModified > directoryLastModified
+    ) {
+      directoryLastModified = file.lastModified;
+    }
+
+    tableElements.push(renderFile(urlPathname, file));
+  }
+
+  const html = handlebarsTemplate({
+    pathname: url.pathname,
+    entries: tableElements,
+  });
+
+  return html;
+}
+
+/**
+ * Element to be displayed on the directory listing page
+ */
+type TableElement = {
+  href: string;
+  name: string;
+  displayNamePaddingRight: string;
+  lastModified: string;
+  size: string;
+};
+
+function renderSubdirectory(name: string): TableElement {
+  const href = encodeURIComponent(name.substring(0, name.length - 1));
+
+  let displayName: string;
+  let displayNamePaddingRight: string;
+  if (name.length > 50) {
+    displayName = name.substring(0, 49) + '>';
+    displayNamePaddingRight = '';
+  } else {
+    displayName = name;
+    displayNamePaddingRight = ' '.repeat(50 - name.length);
+  }
+
+  return {
+    href: `${href}/`,
+    name: displayName,
+    displayNamePaddingRight,
+    lastModified: '               -',
+    size: '                  -',
+  };
+}
+
+function renderFile(pathPrefix: string, file: File): TableElement {
+  const { name, lastModified } = file;
+
+  let displayName: string;
+  let displayNamePaddingRight: string;
+  if (name!.length > 50) {
+    displayName = name.substring(0, 47) + '..>';
+    displayNamePaddingRight = '';
+  } else {
+    displayName = name;
+    displayNamePaddingRight = ' '.repeat(50 - name!.length);
+  }
+
+  const dateString = `${lastModified.getUTCDay()}-${months.at(
```

Comment:
What a mouthful of a string. Let's use `Intl` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DateTimeFormat

Reply:
Trying to get the same output with `Intl`, nginx outputs `29-Apr-2020 22:29` but the closest I could get it to is `21 Jul 2024, 12:26`, not sure if there's any specific locale that would work?

```javascript
new Intl.DateTimeFormat('en-UK', { hourCycle: 'h24', day: 'numeric', month: 'short',  year: 'numeric', hour: '2-digit', minute: '2-digit' }).format(d)
```

Reply:
I guess the closest would be:

```js
const formatter = new Intl.DateTimeFormat('en-GB', {
    day: '2-digit',
    month: 'short',
    year: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    hour12: false
  });
```

And then replace spaces with hyphens? Like `.replace(/(\d{2}) (\w{3}) (\d{4}), (\d{2}):(\d{2})/, '$1-$2-$3 $4:$5');`